// @dosco/minds/build/main/ai/alephalpha.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlephAlpha = exports.AlephAlphaCreativeOptions = exports.AlephAlphaDefaultOptions = exports.AlephaAlphaGenerateHosting = exports.AlephAlphaEmbedRepresentation = exports.AlephAlphaGenerateModel = void 0;
const util_1 = require("./util");
const apiURL = 'https://api.aleph-alpha.com/';
/**
 * AlephAlpha: Models for text generation
 * @export
 */
var AlephAlphaGenerateModel;
(function (AlephAlphaGenerateModel) {
    AlephAlphaGenerateModel["LuminousSupremeControl"] = "luminous-supreme-control";
    AlephAlphaGenerateModel["LuminousSupreme"] = "luminous-supreme";
    AlephAlphaGenerateModel["LuminousExtended"] = "luminous-extended";
    AlephAlphaGenerateModel["LuminousBase"] = "luminous-base";
})(AlephAlphaGenerateModel = exports.AlephAlphaGenerateModel || (exports.AlephAlphaGenerateModel = {}));
var AlephAlphaEmbedRepresentation;
(function (AlephAlphaEmbedRepresentation) {
    AlephAlphaEmbedRepresentation["Symmetric"] = "symmetric";
    AlephAlphaEmbedRepresentation["Document"] = "document";
    AlephAlphaEmbedRepresentation["Query"] = "query";
})(AlephAlphaEmbedRepresentation = exports.AlephAlphaEmbedRepresentation || (exports.AlephAlphaEmbedRepresentation = {}));
var AlephaAlphaGenerateHosting;
(function (AlephaAlphaGenerateHosting) {
    AlephaAlphaGenerateHosting["MaxPrivacy"] = "aleph-alpha";
})(AlephaAlphaGenerateHosting = exports.AlephaAlphaGenerateHosting || (exports.AlephaAlphaGenerateHosting = {}));
/**
 * AlephAlpha: Default Model options for text generation
 * @export
 */
const AlephAlphaDefaultOptions = () => ({
    model: AlephAlphaGenerateModel.LuminousSupreme,
    representation: AlephAlphaEmbedRepresentation.Document,
    disableOptimizations: true,
    maxTokens: 300,
    temperature: 0.45,
    topK: 0,
    topP: 1,
    //frequencyPenalty: 0.40
    // presencePenalty: 0.40
});
exports.AlephAlphaDefaultOptions = AlephAlphaDefaultOptions;
/**
 * AlephAlpha: Default model options for more creative text generation
 * @export
 */
const AlephAlphaCreativeOptions = () => (Object.assign(Object.assign({}, (0, exports.AlephAlphaDefaultOptions)()), { model: AlephAlphaGenerateModel.LuminousSupreme, temperature: 0.9 }));
exports.AlephAlphaCreativeOptions = AlephAlphaCreativeOptions;
const generateReq = (prompt, opt, stopSequences) => ({
    model: opt.model,
    hosting: opt.hosting,
    prompt: prompt,
    maximum_tokens: opt.maxTokens,
    minimum_tokens: opt.minTokens,
    echo: opt.echo,
    temperature: opt.temperature,
    top_k: opt.topK,
    top_p: opt.topP,
    presence_penalty: opt.presencePenalty,
    frequency_penalty: opt.frequencyPenalty,
    sequence_penalty: opt.sequencePenalty,
    sequence_penalty_min_length: opt.sequencePenaltyMinLength,
    repetition_penalties_include_completion: opt.repetitionPenaltiesIncludeCompletion,
    use_multiplicative_presence_penalty: opt.useMultiplicativePresencePenalty,
    useMultiplicativeFrequencyPenalty: opt.useMultiplicativeFrequencyPenalty,
    useMultiplicativeSequencePenalty: opt.useMultiplicativeSequencePenalty,
    penalty_bias: opt.penaltyBias,
    penalty_exceptions: opt.penaltyExceptions,
    penalty_exceptions_include_stop_sequences: opt.penaltyExceptionsIncludeStopSequences,
    best_of: opt.bestOf,
    n: opt.n,
    logit_bias: opt.logitBias,
    log_probs: opt.logProbs,
    stop_sequences: stopSequences,
    tokens: opt.tokens,
    raw_completion: opt.rawCompletion,
    disable_optimizations: opt.disableOptimizations,
    completion_bias_inclusion: opt.completionBiasInclusion,
    completion_bias_inclusion_first_token_only: opt.completionBiasInclusionFirstTokenOnly,
    completion_bias_exclusion: opt.completionBiasExclusion,
    completion_bias_exclusion_first_token_only: opt.completionBiasExclusionFirstTokenOnly,
    contextual_control_threshold: opt.contextualControlThreshold,
    control_log_additive: opt.controlLogAdditive,
});
const embedReq = (prompt, opt) => ({
    model: opt.model,
    hosting: opt.hosting,
    prompt: prompt,
    representation: opt.representation,
    compress_to_size: opt.compressToSize,
    normalize: opt.normalize,
    contextual_control_threshold: opt.contextualControlThreshold,
    control_log_additive: opt.controlLogAdditive,
});
/**
 * AlephAlpha: AI Service
 * @export
 */
class AlephAlpha {
    constructor(apiKey, options = (0, exports.AlephAlphaDefaultOptions)()) {
        if (apiKey === '') {
            throw new Error('AlephAlpha API key not set');
        }
        this.apiKey = apiKey;
        this.options = options;
    }
    name() {
        return 'AlephAlpha';
    }
    generate(prompt, md, sessionID) {
        const text = prompt.trim();
        const res = (0, util_1.apiCall)({
            key: this.apiKey,
            name: "complete" /* apiTypes.Generate */,
            url: apiURL,
        }, generateReq(text, this.options, md === null || md === void 0 ? void 0 : md.stopSequences));
        return res.then(({ completions }) => ({
            id: '',
            sessionID: sessionID,
            query: prompt,
            values: completions.map((v) => ({ id: '', text: v.completion.trim() })),
            value() {
                return this.values[0].text;
            },
        }));
    }
    embed(texts, sessionID) {
        if (texts.length > 1) {
            throw new Error('AlephAlpha limits embeddings input to 1 string');
        }
        const overLimit = texts.filter((v) => v.length > 512);
        if (overLimit.length !== 0) {
            throw new Error('AlephAlpha limits embeddings input to 512 characters');
        }
        const res = (0, util_1.apiCall)({
            key: this.apiKey,
            name: "semantic_embed" /* apiTypes.Embed */,
            url: apiURL,
        }, embedReq(texts[0], this.options));
        return res.then(({ model_version, embedding }) => ({
            id: '',
            sessionID,
            texts,
            model: model_version,
            embeddings: embedding,
        }));
    }
}
exports.AlephAlpha = AlephAlpha;
