// @dosco/minds/build/main/ai/betty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Betty = void 0;
/**
 * Betty: Fake AI Service for writing tests
 * @export
 */
class Betty {
    constructor(answers) {
        this.sdata = new Map();
        this.index = 0;
        this.answers = [...answers];
        this.data = [...answers];
    }
    name() {
        return 'Betty';
    }
    generate(prompt, _md, sessionID) {
        if (sessionID && !this.sdata.has(sessionID)) {
            this.sdata.set(sessionID, [...this.answers]);
        }
        const answers = sessionID ? this.sdata.get(sessionID) : this.data;
        const text = answers.shift();
        this.index++;
        const res = {
            id: this.index.toString(),
            sessionID: sessionID,
            query: prompt,
            values: [{ id: '0', text }],
            value: () => res.values[0].text,
        };
        return new Promise((resolve, _) => {
            setTimeout(() => {
                resolve(res);
            }, 300);
        });
    }
    embed(texts, sessionID) {
        const res = {
            id: '',
            sessionID: sessionID,
            texts: texts,
            model: '',
            embeddings: [1, 2, 3, 4],
        };
        return new Promise((resolve, _) => {
            setTimeout(() => {
                resolve(res);
            }, 300);
        });
    }
}
exports.Betty = Betty;
