// @dosco/minds/build/main/ai/cohere.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cohere = exports.CohereCreativeOptions = exports.CohereDefaultOptions = exports.CohereReturnLikelihoods = exports.CohereGenerateModel = void 0;
const util_1 = require("./util");
const apiURL = 'https://api.cohere.ai/';
/**
 * Cohere: Models for text generation
 * @export
 */
var CohereGenerateModel;
(function (CohereGenerateModel) {
    CohereGenerateModel["CommandXLargeNightly"] = "command-xlarge-nightly";
    CohereGenerateModel["XLarge"] = "xlarge";
    CohereGenerateModel["Medium"] = "medium";
})(CohereGenerateModel = exports.CohereGenerateModel || (exports.CohereGenerateModel = {}));
/**
 * Cohere: Specify how and if the token likelihoods are returned with the response.
 * @export
 */
var CohereReturnLikelihoods;
(function (CohereReturnLikelihoods) {
    CohereReturnLikelihoods["GENERATION"] = "GENERATION";
    CohereReturnLikelihoods["ALL"] = "ALL";
    CohereReturnLikelihoods["NONE"] = "NONE";
})(CohereReturnLikelihoods = exports.CohereReturnLikelihoods || (exports.CohereReturnLikelihoods = {}));
/**
 * Cohere: Default Model options for text generation
 * @export
 */
const CohereDefaultOptions = () => ({
    model: CohereGenerateModel.CommandXLargeNightly,
    maxTokens: 300,
    temperature: 0.45,
    topK: 0,
    topP: 1,
    frequencyPenalty: 0,
    presencePenalty: 0,
});
exports.CohereDefaultOptions = CohereDefaultOptions;
/**
 * Cohere: Default model options for more creative text generation
 * @export
 */
const CohereCreativeOptions = () => (Object.assign(Object.assign({}, (0, exports.CohereDefaultOptions)()), { temperature: 0.9 }));
exports.CohereCreativeOptions = CohereCreativeOptions;
const generateReq = (prompt, opt, stopSequences) => ({
    prompt: prompt,
    model: opt.model,
    max_tokens: opt.maxTokens,
    temperature: opt.temperature,
    k: opt.topK,
    p: opt.topP,
    frequency_penalty: opt.frequencyPenalty,
    presence_penalty: opt.presencePenalty,
    end_sequences: stopSequences,
    stop_sequences: opt.stopSequences,
    return_likelihoods: opt.returnLikelihoods,
});
/**
 * Cohere: AI Service
 * @export
 */
class Cohere {
    constructor(apiKey, options = (0, exports.CohereDefaultOptions)()) {
        if (apiKey === '') {
            throw new Error('Cohere API key not set');
        }
        this.apiKey = apiKey;
        this.options = options;
    }
    name() {
        return 'Cohere';
    }
    generate(prompt, md, sessionID) {
        prompt = prompt.trim();
        const res = (0, util_1.apiCall)({
            key: this.apiKey,
            name: "generate" /* apiTypes.Generate */,
            url: apiURL,
            headers: { 'Cohere-Version': '2022-12-06' },
        }, generateReq(prompt, this.options, md === null || md === void 0 ? void 0 : md.stopSequences));
        return res.then(({ id, generations: gens }) => ({
            id: id,
            sessionID: sessionID,
            query: prompt,
            values: gens,
            value() {
                return this.values[0].text;
            },
        }));
    }
    embed(texts, sessionID) {
        if (texts.length > 96) {
            throw new Error('Cohere limits embeddings input to 96 strings');
        }
        const overLimit = texts.filter((v) => v.length > 512);
        if (overLimit.length !== 0) {
            throw new Error('Cohere limits embeddings input to 512 characters');
        }
        const res = (0, util_1.apiCall)({
            key: this.apiKey,
            name: "embed" /* apiTypes.Embed */,
            url: apiURL,
            headers: { 'Cohere-Version': '2022-12-06' },
        }, { texts, model: this.options.model, truncate: 'NONE' });
        return res.then(({ id, embeddings }) => ({
            id: id,
            sessionID,
            texts,
            model: this.options.model,
            embeddings,
        }));
    }
}
exports.Cohere = Cohere;
