// @dosco/minds/build/module/ai/betty.js
/**
 * Betty: Fake AI Service for writing tests
 * @export
 */
export class Betty {
    answers;
    data;
    sdata = new Map();
    index = 0;
    constructor(answers) {
        this.answers = [...answers];
        this.data = [...answers];
    }
    name() {
        return 'Betty';
    }
    generate(prompt, _md, sessionID) {
        if (sessionID && !this.sdata.has(sessionID)) {
            this.sdata.set(sessionID, [...this.answers]);
        }
        const answers = sessionID ? this.sdata.get(sessionID) : this.data;
        const text = answers.shift();
        this.index++;
        const res = {
            id: this.index.toString(),
            sessionID: sessionID,
            query: prompt,
            values: [{ id: '0', text }],
            value: () => res.values[0].text,
        };
        return new Promise((resolve, _) => {
            setTimeout(() => {
                resolve(res);
            }, 300);
        });
    }
    embed(texts, sessionID) {
        const res = {
            id: '',
            sessionID: sessionID,
            texts: texts,
            model: '',
            embeddings: [1, 2, 3, 4],
        };
        return new Promise((resolve, _) => {
            setTimeout(() => {
                resolve(res);
            }, 300);
        });
    }
}
