// @dosco/minds/build/module/ai/cohere.js
import { apiCall } from './util';
const apiURL = 'https://api.cohere.ai/';
/**
 * Cohere: Models for text generation
 * @export
 */
export var CohereGenerateModel;
(function (CohereGenerateModel) {
    CohereGenerateModel["CommandXLargeNightly"] = "command-xlarge-nightly";
    CohereGenerateModel["XLarge"] = "xlarge";
    CohereGenerateModel["Medium"] = "medium";
})(CohereGenerateModel || (CohereGenerateModel = {}));
/**
 * Cohere: Specify how and if the token likelihoods are returned with the response.
 * @export
 */
export var CohereReturnLikelihoods;
(function (CohereReturnLikelihoods) {
    CohereReturnLikelihoods["GENERATION"] = "GENERATION";
    CohereReturnLikelihoods["ALL"] = "ALL";
    CohereReturnLikelihoods["NONE"] = "NONE";
})(CohereReturnLikelihoods || (CohereReturnLikelihoods = {}));
/**
 * Cohere: Default Model options for text generation
 * @export
 */
export const CohereDefaultOptions = () => ({
    model: CohereGenerateModel.CommandXLargeNightly,
    maxTokens: 300,
    temperature: 0.45,
    topK: 0,
    topP: 1,
    frequencyPenalty: 0,
    presencePenalty: 0,
});
/**
 * Cohere: Default model options for more creative text generation
 * @export
 */
export const CohereCreativeOptions = () => ({
    ...CohereDefaultOptions(),
    temperature: 0.9,
});
const generateReq = (prompt, opt, stopSequences) => ({
    prompt: prompt,
    model: opt.model,
    max_tokens: opt.maxTokens,
    temperature: opt.temperature,
    k: opt.topK,
    p: opt.topP,
    frequency_penalty: opt.frequencyPenalty,
    presence_penalty: opt.presencePenalty,
    end_sequences: stopSequences,
    stop_sequences: opt.stopSequences,
    return_likelihoods: opt.returnLikelihoods,
});
/**
 * Cohere: AI Service
 * @export
 */
export class Cohere {
    apiKey;
    options;
    constructor(apiKey, options = CohereDefaultOptions()) {
        if (apiKey === '') {
            throw new Error('Cohere API key not set');
        }
        this.apiKey = apiKey;
        this.options = options;
    }
    name() {
        return 'Cohere';
    }
    generate(prompt, md, sessionID) {
        prompt = prompt.trim();
        const res = apiCall({
            key: this.apiKey,
            name: "generate" /* apiTypes.Generate */,
            url: apiURL,
            headers: { 'Cohere-Version': '2022-12-06' },
        }, generateReq(prompt, this.options, md?.stopSequences));
        return res.then(({ id, generations: gens }) => ({
            id: id,
            sessionID: sessionID,
            query: prompt,
            values: gens,
            value() {
                return this.values[0].text;
            },
        }));
    }
    embed(texts, sessionID) {
        if (texts.length > 96) {
            throw new Error('Cohere limits embeddings input to 96 strings');
        }
        const overLimit = texts.filter((v) => v.length > 512);
        if (overLimit.length !== 0) {
            throw new Error('Cohere limits embeddings input to 512 characters');
        }
        const res = apiCall({
            key: this.apiKey,
            name: "embed" /* apiTypes.Embed */,
            url: apiURL,
            headers: { 'Cohere-Version': '2022-12-06' },
        }, { texts, model: this.options.model, truncate: 'NONE' });
        return res.then(({ id, embeddings }) => ({
            id: id,
            sessionID,
            texts,
            model: this.options.model,
            embeddings,
        }));
    }
}
