// @dosco/minds/build/main/lib/useai.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cohere = void 0;
const axios_1 = __importDefault(require("axios"));
const defaultOptions = {
    version: '2022-12-06',
    model: 'command-xlarge-nightly',
    maxTokens: 300,
    temperature: 0.9,
    k: 0,
    p: 0.75,
    frequencyPenalty: 0,
    presencePenalty: 0,
    stopSequences: [],
    returnLikelihoods: 'NONE',
};
const cohereQuery = async (prompt, apiKey, opt) => {
    const options = {
        headers: {
            Authorization: `BEARER ${apiKey}`,
            'Cohere-Version': opt.version,
        },
    };
    const data = {
        prompt: prompt,
        model: opt.model,
        max_tokens: opt.maxTokens,
        temperature: opt.temperature,
        k: opt.k,
        p: opt.p,
        frequency_penalty: opt.frequencyPenalty,
        presence_penalty: opt.presencePenalty,
        stop_sequences: opt.stopSequences,
        return_likelihoods: opt.returnLikelihoods,
    };
    const res = await axios_1.default.post('https://api.cohere.ai/generate', data, options);
    return res.data;
};
const Cohere = (apiKey, opt = defaultOptions) => {
    return {
        query: (prompt) => cohereQuery(prompt, apiKey, opt),
    };
};
exports.Cohere = Cohere;
