// @dosco/minds/build/main/text/useai.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateText = void 0;
class GenerateText {
    constructor(ai, mem, prompt) {
        this.maxSteps = 20;
        this.debug = false;
        this.ai = ai;
        this.mem = mem;
        this.prompt = prompt;
        if (prompt.actions) {
            this.actmap = prompt
                .actions()
                .reduce((m, a) => (m.set(a.name, a), m), new Map());
        }
    }
    setMaxSteps(n) {
        this.maxSteps = n;
    }
    setDebug(b) {
        this.debug = b;
    }
    generate(query) {
        return new Promise((resolve) => {
            const res = this._generate(query);
            resolve(res);
        });
    }
    async _generate(query) {
        const md = this.prompt.metadata();
        const q = query.trim();
        if (q === '') {
            throw new Error('not query found');
        }
        for (let i = 0; i < this.maxSteps; i++) {
            const res = await this.callAI(q, md);
            let done = false;
            if (this.debug) {
                log(`< ${res.value}`, 'lime');
            }
            if (this.prompt.actions) {
                done = this.processAction(res, md);
            }
            else {
                done = this.processSeq(q, res, md);
            }
            if (res.value === '') {
                throw new Error('empty response from ai');
            }
            if (done) {
                return res;
            }
        }
        throw new Error(`query uses over max number of steps: ${this.maxSteps}`);
    }
    callAI(query, md) {
        const p = this.prompt.create(this.ai, this.mem, query);
        if (this.debug) {
            log(`> ${p}`, 'white');
        }
        return this.ai.generate(p, md);
    }
    processSeq(query, res, md) {
        const val = [md === null || md === void 0 ? void 0 : md.queryPrefix, query, md === null || md === void 0 ? void 0 : md.responsePrefix, res.value];
        this.mem.add(val.join(''));
        return true;
    }
    processAction(res, md) {
        if (!md.actionName) {
            throw new Error('actionName parameter not set');
        }
        if (!md.actionValue) {
            throw new Error('actionValue parameter not set');
        }
        if (!md.finalValue) {
            throw new Error('finalValue parameter not set');
        }
        let actKey;
        let actVal;
        let v;
        const val = res.value;
        if ((v = md.finalValue.exec(val)) !== null) {
            const mval = [md === null || md === void 0 ? void 0 : md.responsePrefix, val];
            this.mem.add(mval.join(''));
            res.value = v[1].trim();
            return true;
        }
        if ((v = md.actionName.exec(res.value)) !== null) {
            actKey = v[1].trim();
        }
        if ((v = md.actionValue.exec(res.value)) !== null) {
            actVal = v[1].trim();
        }
        if (!this.actmap.has(actKey)) {
            throw new Error(`invalid action found: ${actKey}`);
        }
        const act = this.actmap.get(actKey);
        const actionResult = act.action(actVal);
        if (this.debug) {
            log(`> ${actKey}(${actVal}): ${actionResult}`, 'cyan');
        }
        const mval = [md === null || md === void 0 ? void 0 : md.responsePrefix, val, md === null || md === void 0 ? void 0 : md.queryPrefix, actionResult];
        this.mem.add(mval.join(''));
        return false;
    }
}
exports.GenerateText = GenerateText;
function log(msg, color) {
    // @ts-ignore
    if (typeof window === 'undefined') {
        if (color === 'lime') {
            color = '\u001b[32;1m';
        }
        if (color === 'cyan') {
            color = '\x1b[36m';
        }
        if (color === 'white') {
            color = '\x1b[1m\u001b[37m';
        }
        console.log(`${color}${msg}\x1b[0m\n`);
    }
    else {
        console.log(`%c${msg}`, { color });
    }
}
/*
Magic isn't always unicorns and fairy dust␊
Sometimes it's computer code and technology that's quite a bit of work␊
But if you learn the tricks and the trades␊
You'll find there's a wonder that awaits␊

With a computer and some basic coding␊
You can make anything your heart desires␊
From games that you'll play for hours on end␊
To apps that will make life much more fun␊

So don't be afraid of the unknown␊
Embrace the magic of computer code␊
And you'll find that your dreams can come true␊
With just a little bit of coding, you can do anything too!
*/
